Redesign checksum-related timeouts and reports

marciobarbosa opened the pull request:
Currently we have a timeout of 1800 seconds (30 minutes) to
calculate the checksum of a volume dump. On average, this timeout
should be enough in most of the cases. However, a single timeout
for the entire file is not the best approach for a couple of reasons.
First of all, the user does not get any report about the current
progress of this state. As a result, he has no idea if CellCC is stuck
or making some progress. Secondly, this timeout might still not be
enough for the whole file.

In order to fix the problems reported above, a new function was
designed to calculate the checksum of the whole dump file. This
function breaks the file into smaller chunks, computes the checksum
of these smaller pieces, and add them all in separated steps. Also,
this function reports the progress of this state minute by minute.
Last, but not least, the timeout of the state in question is updated
every single minute.

adeason commented on lib/AFS/CellCC/Util.pm:
```diff
@@ -331,4 +332,58 @@ scratch_ok($$$$$) {
+    }
+
+    my $start = time();
+    my $now = $start;
```
This assignment seems pointless. Just declare 'my $now;'

adeason commented on lib/AFS/CellCC/Util.pm:
```diff
@@ -331,4 +332,58 @@ scratch_ok($$$$$) {
+# Return the checksum of the file received as an argument.
+# The checksum returned by this function will be in following
+# format: "algorithm:checksum", where algorithm is the algorithm
+# of choice (e.g. MD5) and checksum is the checksum of the file
```
Could you mention that the algorithm is one that is recognized by the Digest module? Just to be more specific about what we mean by the algorithm name.

adeason commented on lib/AFS/CellCC/Util.pm:
```diff
@@ -331,4 +332,58 @@ scratch_ok($$$$$) {
+    my $bytes;
+    my $nbytes;
+
+    my $digest = Digest->new($algo);
```
Since this module is now using the Digest module, add a 'use Digest;' near the beginning of this file. Currently I believe this doesn't cause an error, because the only other places in the code (Dump.pm and Restore.pm) of course already 'use Digest'. Also remove the 'use Digest' from Dump.pm (but not Restore.pm, since Restore.pm still uses Digest directly).

adeason commented on lib/AFS/CellCC/Util.pm:
```diff
@@ -331,4 +332,58 @@ scratch_ok($$$$$) {
+
+        $now = time();
+        if ($now < $start || $now - $start > 60) {
+            $bytes = pretty_bytes(tell($dumpfh));
```
Looking at this now, I'd rather have just a little more error checking. Something like:

my $pos = tell($dumpfh);
if ($pos < 0) {
    die("'tell' failed: $!\n");
}
$bytes = pretty_bytes($pos);

adeason commented:
Looks good. If you can fix the minor things I mentioned in the diff notes, I'll merge it.

marciobarbosa committed:
Redesign checksum-related timeouts and reports

Currently we have a timeout of 1800 seconds (30 minutes) to
calculate the checksum of a volume dump. On average, this timeout
should be enough in most of the cases. However, a single timeout
for the entire file is not the best approach for a couple of reasons.
First of all, the user does not get any report about the current
progress of this state. As a result, he has no idea if CellCC is stuck
or making some progress. Secondly, this timeout might still not be
enough for the whole file.

In order to fix the problems reported above, a new function was
designed to calculate the checksum of the whole dump file. This
function breaks the file into smaller chunks, computes the checksum
of these smaller pieces, and add them all in separated steps. Also,
this function reports the progress of this state minute by minute.
Last, but not least, the timeout of the state in question is updated
every single minute.
